File: 0129-sum-root-to-leaf-numbers/0129-sum-root-to-leaf-numbers.py
class Solution:
    def sumNumbers(self, root: Optional[TreeNode]) -> int:
        def calc(node, total):
            if node.left is None and node.right is None:
                nonlocal result
                result += total
                return
            if node.left:
                calc(node.left, total * 10 + node.left.val)
            if node.right:
                calc(node.right, total * 10 + node.right.val)


        if not root: return 0
        result = 0
        calc(root, root.val)
        return result

File: 0129-sum-root-to-leaf-numbers/0129-sum-root-to-leaf-numbers.java
class Solution {
    private int result = 0;

    public int sumNumbers(TreeNode root) {
        if (root == null) {
            return 0;
        }
        calc(root, root.val);
        return result;
    }

    private void calc(TreeNode node, int total) {
        if (node.left == null && node.right == null) {
            result += total;
            return;
        }
        if (node.left != null) {
            calc(node.left, total * 10 + node.left.val);
        }
        if (node.right != null) {
            calc(node.right, total * 10 + node.right.val);
        }
    }
}